// mongodb/src/read_concern.ts
import type { Document } from './bson';


export const ReadConcernLevel = Object.freeze({
  local: 'local',
  majority: 'majority',
  linearizable: 'linearizable',
  available: 'available',
  snapshot: 'snapshot'
} as const);


export type ReadConcernLevel = (typeof ReadConcernLevel)[keyof typeof ReadConcernLevel];


export type ReadConcernLike = ReadConcern | { level: ReadConcernLevel } | ReadConcernLevel;


export class ReadConcern {
  level: ReadConcernLevel | string;

  
  constructor(level: ReadConcernLevel) {
    
    this.level = ReadConcernLevel[level] ?? level;
  }

  
  static fromOptions(options?: {
    readConcern?: ReadConcernLike;
    level?: ReadConcernLevel;
  }): ReadConcern | undefined {
    if (options == null) {
      return;
    }

    if (options.readConcern) {
      const { readConcern } = options;
      if (readConcern instanceof ReadConcern) {
        return readConcern;
      } else if (typeof readConcern === 'string') {
        return new ReadConcern(readConcern);
      } else if ('level' in readConcern && readConcern.level) {
        return new ReadConcern(readConcern.level);
      }
    }

    if (options.level) {
      return new ReadConcern(options.level);
    }
    return;
  }

  static get MAJORITY(): 'majority' {
    return ReadConcernLevel.majority;
  }

  static get AVAILABLE(): 'available' {
    return ReadConcernLevel.available;
  }

  static get LINEARIZABLE(): 'linearizable' {
    return ReadConcernLevel.linearizable;
  }

  static get SNAPSHOT(): 'snapshot' {
    return ReadConcernLevel.snapshot;
  }

  toJSON(): Document {
    return { level: this.level };
  }
}
